sound/soc/sof/topology.c

```diff
@@ -1957,21 +1963,37 @@ static int sof_widget_unload(struct snd_soc_component *scomp,
 		/* update enabled cores mask */
 		sdev->enabled_cores_mask &= ~(1 << pipeline->core);
 
-		break;
-	case snd_soc_dapm_pga:
-
-		/* get volume kcontrol */
-		kc = &widget->kcontrol_news[0];
-		sm = (struct soc_mixer_control *)kc->private_value;
-		scontrol = sm->dobj.private;
-
-		/* free volume table */
-		kfree(scontrol->volume_table);
 		break;
 	default:
 		break;
 	}
+	for (i = 0; i < widget->num_kcontrols; i++) {
+		kc = &widget->kcontrol_news[i];
+		switch (dobj->widget.kcontrol_type) {
+		case SND_SOC_TPLG_TYPE_MIXER:
+			sm = (struct soc_mixer_control *)kc->private_value;
+			scontrol = sm->dobj.private;
+			if (sm->max > 1)
+				kfree(scontrol->volume_table);
+			break;
+		case SND_SOC_TPLG_TYPE_ENUM:
+			se = (struct soc_enum *)kc->private_value;
+			scontrol = se->dobj.private;
+			break;
+		case SND_SOC_TPLG_TYPE_BYTES:
+			sbe = (struct soc_bytes_ext *)kc->private_value;
+			scontrol = sbe->dobj.private;
+			break;
+		default:
+			dev_warn(sdev->dev, "unsupported kcontrol_tupe\n");
```

Comment:
@bardliao typo here. kcontrol type?

Reply:
@ranj063 Thanks. I also update the goto label to lower case to be consistent with others.